Geoshape properties can't be deserialized in OLAP mode

A KryoException is thrown when attempting to deserialize any Geoshape properties in OLAP mode because Geoshape is missing a no-arg constructor.

```
Caused by: org.apache.spark.SparkException: Job aborted due to stage failure: Exception while getting task result: java.io.IOException: org.apache.tinkerpop.shaded.kryo.KryoException: Class cannot be created (missing no-arg constructor): org.janusgraph.core.attribute.Geos
hape
Serialization trace:
value (org.apache.tinkerpop.gremlin.structure.util.detached.DetachedVertexProperty)
properties (org.apache.tinkerpop.gremlin.structure.util.detached.DetachedVertex)
t (org.apache.tinkerpop.gremlin.process.traversal.traverser.B_O_Traverser)
        at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1431)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1419)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1418)
        at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
        at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:47)
        at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1418)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
        at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:799)
        at scala.Option.foreach(Option.scala:236)
        at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:799)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1640)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1599)
        at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1588)
        at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
        at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:620)
```

--- a/janusgraph-all/src/main/resources/META-INF/services/org.apache.tinkerpop.gremlin.structure.io.gryo.kryoshim.KryoShimService
+++ b/janusgraph-all/src/main/resources/META-INF/services/org.apache.tinkerpop.gremlin.structure.io.gryo.kryoshim.KryoShimService
@@ -1,0 +1,1 @@
+org.janusgraph.hadoop.serialize.JanusGraphKryoShimService

--- a/janusgraph-dist/src/assembly/static/bin/gremlin-server.bat
+++ b/janusgraph-dist/src/assembly/static/bin/gremlin-server.bat
@@ -88,7 +88,8 @@ IF NOT DEFINED JAVA_OPTIONS (
  -Dlog4j.configuration=file:/%JANUSGRAPH_HOME%\conf\gremlin-server\log4j-server.properties ^
  -Dlog4j.debug=true ^
  -Dgremlin.log4j.level=%GREMLIN_LOG_LEVEL% ^
- -javaagent:%JAMM_JAR%
+ -javaagent:%JAMM_JAR% ^
+ -Dgremlin.io.kryoShimService=org.janusgraph.hadoop.serialize.JanusGraphKryoShimService
 )
 
 

--- a/janusgraph-dist/src/assembly/static/bin/gremlin-server.sh
+++ b/janusgraph-dist/src/assembly/static/bin/gremlin-server.sh
@@ -50,7 +50,7 @@ fi
 
 # Set Java options
 if [ "$JAVA_OPTIONS" = "" ] ; then
-    JAVA_OPTIONS="-Xms32m -Xmx512m -javaagent:$LIB/jamm-0.3.0.jar"
+    JAVA_OPTIONS="-Xms32m -Xmx512m -javaagent:$LIB/jamm-0.3.0.jar -Dgremlin.io.kryoShimService=org.janusgraph.hadoop.serialize.JanusGraphKryoShimService"
 fi
 
 # Execute the application and return its exit code

--- a/janusgraph-dist/src/assembly/static/bin/gremlin.bat
+++ b/janusgraph-dist/src/assembly/static/bin/gremlin.bat
@@ -88,7 +88,8 @@ IF NOT DEFINED JAVA_OPTIONS (
  -Dlog4j.configuration=file:/%JANUSGRAPH_HOME%\conf\log4j-console.properties ^
  -Dgremlin.log4j.level=%GREMLIN_LOG_LEVEL% ^
  -Djline.terminal=none ^
- -javaagent:%JAMM_JAR%
+ -javaagent:%JAMM_JAR% ^
+ -Dgremlin.io.kryoShimService=org.janusgraph.hadoop.serialize.JanusGraphKryoShimService
 )
 
 :: Launch the application

--- a/janusgraph-dist/src/assembly/static/bin/gremlin.sh
+++ b/janusgraph-dist/src/assembly/static/bin/gremlin.sh
@@ -119,7 +119,7 @@ if [ -z "${HADOOP_GREMLIN_LIBS:-}" ]; then
     export HADOOP_GREMLIN_LIBS="$LIB"
 fi
 
-JAVA_OPTIONS="${JAVA_OPTIONS} -Duser.working_dir=${USER_DIR} -Dtinkerpop.ext=${USER_EXT_DIR:-${SYSTEM_EXT_DIR}} -Dlog4j.configuration=conf/log4j-console.properties -Dgremlin.log4j.level=$GREMLIN_LOG_LEVEL -javaagent:$LIB/jamm-0.3.0.jar"
+JAVA_OPTIONS="${JAVA_OPTIONS} -Duser.working_dir=${USER_DIR} -Dtinkerpop.ext=${USER_EXT_DIR:-${SYSTEM_EXT_DIR}} -Dlog4j.configuration=conf/log4j-console.properties -Dgremlin.log4j.level=$GREMLIN_LOG_LEVEL -javaagent:$LIB/jamm-0.3.0.jar -Dgremlin.io.kryoShimService=org.janusgraph.hadoop.serialize.JanusGraphKryoShimService"
 JAVA_OPTIONS=$(awk -v RS=' ' '!/^$/ {if (!x[$0]++) print}' <<< "${JAVA_OPTIONS}" | grep -v '^$' | paste -sd ' ' -)
 
 if [ -n "$SCRIPT_DEBUG" ]; then

--- a/janusgraph-dist/src/pkg/static/pkgcommon/payload/usr/bin/gremlin.sh
+++ b/janusgraph-dist/src/pkg/static/pkgcommon/payload/usr/bin/gremlin.sh
@@ -119,7 +119,7 @@ if [ -z "${HADOOP_GREMLIN_LIBS:-}" ]; then
     export HADOOP_GREMLIN_LIBS="$LIB"
 fi
 
-JAVA_OPTIONS="${JAVA_OPTIONS} -Duser.working_dir=${USER_DIR} -Dtinkerpop.ext=${USER_EXT_DIR:-${SYSTEM_EXT_DIR}} -Dlog4j.configuration=conf/log4j-console.properties -Dgremlin.log4j.level=$GREMLIN_LOG_LEVEL -javaagent:$LIB/jamm-0.3.0.jar"
+JAVA_OPTIONS="${JAVA_OPTIONS} -Duser.working_dir=${USER_DIR} -Dtinkerpop.ext=${USER_EXT_DIR:-${SYSTEM_EXT_DIR}} -Dlog4j.configuration=conf/log4j-console.properties -Dgremlin.log4j.level=$GREMLIN_LOG_LEVEL -javaagent:$LIB/jamm-0.3.0.jar -Dgremlin.io.kryoShimService=org.janusgraph.hadoop.serialize.JanusGraphKryoShimService"
 JAVA_OPTIONS=$(awk -v RS=' ' '!/^$/ {if (!x[$0]++) print}' <<< "${JAVA_OPTIONS}" | grep -v '^$' | paste -sd ' ' -)
 
 if [ -n "$SCRIPT_DEBUG" ]; then

--- a/janusgraph-hadoop-parent/janusgraph-hadoop-core/src/main/java/org/janusgraph/hadoop/serialize/JanusGraphKryoShimService.java
+++ b/janusgraph-hadoop-parent/janusgraph-hadoop-core/src/main/java/org/janusgraph/hadoop/serialize/JanusGraphKryoShimService.java
@@ -1,0 +1,18 @@
+package org.janusgraph.hadoop.serialize;
+
+import com.google.common.collect.ImmutableList;
+import org.apache.commons.configuration.BaseConfiguration;
+import org.apache.tinkerpop.gremlin.hadoop.structure.io.HadoopPoolShimService;
+import org.apache.tinkerpop.gremlin.hadoop.structure.io.HadoopPools;
+import org.apache.tinkerpop.gremlin.structure.io.gryo.GryoPool;
+import org.janusgraph.graphdb.tinkerpop.JanusGraphIoRegistry;
+
+public class JanusGraphKryoShimService extends HadoopPoolShimService {
+
+    public JanusGraphKryoShimService() {
+        final BaseConfiguration c = new BaseConfiguration();
+        c.setProperty(GryoPool.CONFIG_IO_REGISTRY, ImmutableList.of(JanusGraphIoRegistry.class.getCanonicalName()));
+        HadoopPools.initialize(c);
+    }
+
+}

--- a/janusgraph-hadoop-parent/janusgraph-hadoop-core/src/test/java/org/janusgraph/hadoop/AbstractInputFormatIT.java
+++ b/janusgraph-hadoop-parent/janusgraph-hadoop-core/src/test/java/org/janusgraph/hadoop/AbstractInputFormatIT.java
@@ -106,5 +106,19 @@ public abstract class AbstractInputFormatIT extends JanusGraphBaseTest {
         assertEquals(2, edges.size());
     }
 
+    @Test
+    public void testGeoshapeGetValues() throws Exception {
+        GraphOfTheGodsFactory.load(graph, null, true);
+
+        // Read geoshape using the inputformat
+        Graph g = getGraph();
+        GraphTraversalSource t = g.traversal(GraphTraversalSource.computer(SparkGraphComputer.class));
+        Iterator<Object> geoIter = t.E().values("place");
+        assertNotNull(geoIter);
+        assertTrue(geoIter.hasNext());
+        Set<Object> geos = Sets.newHashSet(geoIter);
+        assertEquals(3, geos.size());
+    }
+
     abstract protected Graph getGraph() throws IOException, ConfigurationException;
 }

--- a/janusgraph-hadoop-parent/janusgraph-hadoop-core/src/test/resources/META-INF/services/org.apache.tinkerpop.gremlin.structure.io.gryo.kryoshim.KryoShimService
+++ b/janusgraph-hadoop-parent/janusgraph-hadoop-core/src/test/resources/META-INF/services/org.apache.tinkerpop.gremlin.structure.io.gryo.kryoshim.KryoShimService
@@ -1,0 +1,1 @@
+org.janusgraph.hadoop.serialize.JanusGraphKryoShimService

